# JSON examples and exercise
****
+ get familiar with packages for dealing with JSON
+ study examples with JSON strings and files 
+ work on exercise to be completed and submitted 
****
+ reference: http://pandas.pydata.org/pandas-docs/stable/io.html#io-json-reader
+ data source: http://jsonstudio.com/resources/
****

In [1]:
import pandas as pd

## imports for Python, Pandas

In [2]:
import json
from pandas.io.json import json_normalize

## JSON example, with string

+ demonstrates creation of normalized dataframes (tables) from nested json string
+ source: http://pandas.pydata.org/pandas-docs/stable/io.html#normalization

In [3]:
# define json string
data = [{'state': 'Florida', 
         'shortname': 'FL',
         'info': {'governor': 'Rick Scott'},
         'counties': [{'name': 'Dade', 'population': 12345},
                      {'name': 'Broward', 'population': 40000},
                      {'name': 'Palm Beach', 'population': 60000}]},
        {'state': 'Ohio',
         'shortname': 'OH',
         'info': {'governor': 'John Kasich'},
         'counties': [{'name': 'Summit', 'population': 1234},
                      {'name': 'Cuyahoga', 'population': 1337}]}]

In [4]:
# use normalization to create tables from nested element
json_normalize(data, 'counties')

name  population
0        Dade       12345
1     Broward       40000
2  Palm Beach       60000
3      Summit        1234
4    Cuyahoga        1337

In [5]:
# further populate tables created from nested element
json_normalize(data, 'counties', ['state', 'shortname', ['info', 'governor']])

name  population    state shortname info.governor
0        Dade       12345  Florida        FL    Rick Scott
1     Broward       40000  Florida        FL    Rick Scott
2  Palm Beach       60000  Florida        FL    Rick Scott
3      Summit        1234     Ohio        OH   John Kasich
4    Cuyahoga        1337     Ohio        OH   John Kasich

****
## JSON example, with file

+ demonstrates reading in a json file as a string and as a table
+ uses small sample file containing data about projects funded by the World Bank 
+ data source: http://jsonstudio.com/resources/

In [6]:
# load json as string
json.load((open('data/world_bank_projects_less.json')))

[{'_id': {'$oid': '52b213b38594d8a2be17c780'},
  'approvalfy': 1999,
  'board_approval_month': 'November',
  'boardapprovaldate': '2013-11-12T00:00:00Z',
  'borrower': 'FEDERAL DEMOCRATIC REPUBLIC OF ETHIOPIA',
  'closingdate': '2018-07-07T00:00:00Z',
  'country_namecode': 'Federal Democratic Republic of Ethiopia!$!ET',
  'countrycode': 'ET',
  'countryname': 'Federal Democratic Republic of Ethiopia',
  'countryshortname': 'Ethiopia',
  'docty': 'Project Information Document,Indigenous Peoples Plan,Project Information Document',
  'envassesmentcategorycode': 'C',
  'grantamt': 0,
  'ibrdcommamt': 0,
  'id': 'P129828',
  'idacommamt': 130000000,
  'impagency': 'MINISTRY OF EDUCATION',
  'lendinginstr': 'Investment Project Financing',
  'lendinginstrtype': 'IN',
  'lendprojectcost': 550000000,
  'majorsector_percent': [{'Name': 'Education', 'Percent': 46},
   {'Name': 'Education', 'Percent': 26},
   {'Name': 'Public Administration, Law, and Justice', 'Percent': 16},
   {'Name': 'Educatio

In [7]:
# load as Pandas dataframe
sample_json_df = pd.read_json('data/world_bank_projects_less.json')
sample_json_df

_id  approvalfy board_approval_month  \
0  {'$oid': '52b213b38594d8a2be17c780'}        1999             November   
1  {'$oid': '52b213b38594d8a2be17c781'}        2015             November   

      boardapprovaldate                                 borrower  \
0  2013-11-12T00:00:00Z  FEDERAL DEMOCRATIC REPUBLIC OF ETHIOPIA   
1  2013-11-04T00:00:00Z                    GOVERNMENT OF TUNISIA   

            closingdate                              country_namecode  \
0  2018-07-07T00:00:00Z  Federal Democratic Republic of Ethiopia!$!ET   
1                   NaN                      Republic of Tunisia!$!TN   

  countrycode                              countryname countryshortname  \
0          ET  Federal Democratic Republic of Ethiopia         Ethiopia   
1          TN                      Republic of Tunisia          Tunisia   

                         ...                           sectorcode source  \
0                        ...                          ET,BS,ES,EP   IBRD   
1                        ...                                BZ,BS   IBRD   

   status  supplementprojectflg  \
0  Active                     N   
1  Active                     N   

                                              theme1  \
0      {'Name': 'Education for all', 'Percent': 100}   
1  {'Name': 'Other economic management', 'Percent...   

                                      theme_namecode themecode   totalamt  \
0      [{'name': 'Education for all', 'code': '65'}]        65  130000000   
1  [{'name': 'Other economic management', 'code':...     54,24          0   

  totalcommamt                                                url  
0    130000000  http://www.worldbank.org/projects/P129828/ethi...  
1      4700000  http://www.worldbank.org/projects/P144674?lang=en  

[2 rows x 50 columns]

****
## JSON exercise

Using data in file 'data/world_bank_projects.json' and the techniques demonstrated above,
1. Find the 10 countries with most projects
2. Find the top 10 major project themes (using column 'mjtheme_namecode')
3. In 2. above you will notice that some entries have only the code and the name is missing. Create a dataframe with the missing names filled in.

In [8]:
sample_json_df.columns

Index(['_id', 'approvalfy', 'board_approval_month', 'boardapprovaldate',
       'borrower', 'closingdate', 'country_namecode', 'countrycode',
       'countryname', 'countryshortname', 'docty', 'envassesmentcategorycode',
       'grantamt', 'ibrdcommamt', 'id', 'idacommamt', 'impagency',
       'lendinginstr', 'lendinginstrtype', 'lendprojectcost',
       'majorsector_percent', 'mjsector_namecode', 'mjtheme',
       'mjtheme_namecode', 'mjthemecode', 'prodline', 'prodlinetext',
       'productlinetype', 'project_abstract', 'project_name', 'projectdocs',
       'projectfinancialtype', 'projectstatusdisplay', 'regionname', 'sector',
       'sector1', 'sector2', 'sector3', 'sector4', 'sector_namecode',
       'sectorcode', 'source', 'status', 'supplementprojectflg', 'theme1',
       'theme_namecode', 'themecode', 'totalamt', 'totalcommamt', 'url'],
      dtype='object')

### 1. Find the 10 countries with most projects

In [9]:
#Set data path
path = 'data/world_bank_projects.json'
#load raw data
raw = json.load(open('data/world_bank_projects.json'))
#load as data frame
raw_data = pd.read_json('data/world_bank_projects.json')

In [10]:
#Create data frame with countries and project status
data_1 = pd.DataFrame({'Country': raw_data.countryname, 'Status': raw_data.status})
print('Possible values for Status column:', data_1.Status.unique())

Possible values for Status column: ['Active' 'Closed']


The column Status has only two possible values, so its type can be set to category type.

In [11]:
data_1.Status = data_1.Status.astype('category')

In [12]:
print(data_1.head())

                                   Country  Status
0  Federal Democratic Republic of Ethiopia  Active
1                      Republic of Tunisia  Active
2                                   Tuvalu  Active
3                        Republic of Yemen  Active
4                       Kingdom of Lesotho  Active


In [13]:
#Information on data_1
print(data_1.info())

<class 'pandas.core.frame.DataFrame'>
Int64Index: 500 entries, 0 to 499
Data columns (total 2 columns):
Country    500 non-null object
Status     500 non-null category
dtypes: category(1), object(1)
memory usage: 8.4+ KB
None


There are no null entries in data_1.
We add a column to data_1 called Value initialized with zeros. When a project is active, the Value column changes to 1 and stays at 0 for closed projects.

In [14]:
data_1['Value'] = 0
data_1.Value[data_1.Status == 'Active'] = 1

/home/pfgr/anaconda3/envs/py3/lib/python3.6/site-packages/ipykernel_launcher.py:2: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/indexing.html#indexing-view-versus-copy
  


We pivot data_1 in such a way that the countries become indeces, the 'Status' categories become columns and we use as an aggregate function 'count' to get the total number closed projects and active projects per country.

In [15]:
pivot_data = data_1.pivot_table(index = 'Country', columns='Status', values = 'Value', aggfunc = 'count')

In [16]:
print(pivot_data.info())

<class 'pandas.core.frame.DataFrame'>
Index: 118 entries, Africa to World
Data columns (total 2 columns):
Active    112 non-null float64
Closed    44 non-null float64
dtypes: float64(2)
memory usage: 2.8+ KB
None


There are null values in the 'Closed' column, so by assuming that a null value corresponds to zero closed projects,  we replace null values by zeros.

In [17]:
pivot_data.fillna(0, inplace=True)

In [18]:
pivot_data.info()

<class 'pandas.core.frame.DataFrame'>
Index: 118 entries, Africa to World
Data columns (total 2 columns):
Active    118 non-null float64
Closed    118 non-null float64
dtypes: float64(2)
memory usage: 2.8+ KB


data_1 has 500 entries. We want to check that the number of active projects plus the number of closed projects is also 500.

In [19]:
print('The number of active projects plus the number of closed projects is:', \
      pivot_data.Active.sum() + pivot_data.Closed.sum())

The number of active projects plus the number of closed projects is: 500.0


Given that the total number of projects match the number of entries of the initial data, we can continue by adding a new column to the pivot_data capturing the total number of projects (meaning the sum of closed plus active projects per country).

In [20]:
pivot_data = pd.merge(pd.DataFrame(pivot_data['Active'] + pivot_data['Closed'], \
                                   columns=['Total']), pivot_data, left_index=True, right_index=True)

In [21]:
print('The 10 countries with most projects (including closed and active projects) are:\n\n', \
      pivot_data.sort_values(by=['Total'], ascending = False).head(10))

The 10 countries with most projects (including closed and active projects) are:

                                  Total  Active  Closed
Country                                               
People's Republic of China        19.0    19.0     0.0
Republic of Indonesia             19.0    16.0     3.0
Socialist Republic of Vietnam     17.0    14.0     3.0
Republic of India                 16.0    15.0     1.0
Republic of Yemen                 13.0    13.0     0.0
People's Republic of Bangladesh   12.0    12.0     0.0
Nepal                             12.0    11.0     1.0
Kingdom of Morocco                12.0     6.0     6.0
Republic of Mozambique            11.0     9.0     2.0
Africa                            11.0    11.0     0.0


In [22]:
print('However, the 10 countries with most active projects are:\n\n', \
      pivot_data.sort_values(by=['Active'], ascending = False).head(10))

However, the 10 countries with most active projects are:

                                  Total  Active  Closed
Country                                               
People's Republic of China        19.0    19.0     0.0
Republic of Indonesia             19.0    16.0     3.0
Republic of India                 16.0    15.0     1.0
Socialist Republic of Vietnam     17.0    14.0     3.0
Republic of Yemen                 13.0    13.0     0.0
People's Republic of Bangladesh   12.0    12.0     0.0
Africa                            11.0    11.0     0.0
Nepal                             12.0    11.0     1.0
Islamic Republic of Pakistan       9.0     9.0     0.0
Republic of Mozambique            11.0     9.0     2.0


### 2. Find the top 10 major project themes (using column 'mjtheme_namecode')

In [23]:
#Data frame with major themes
themes_code = json_normalize(raw, 'mjtheme_namecode')

In [24]:
print(themes_code.info())

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1499 entries, 0 to 1498
Data columns (total 2 columns):
code    1499 non-null object
name    1499 non-null object
dtypes: object(2)
memory usage: 23.5+ KB
None


No null values present.

In [25]:
themes_code.head()

code                                   name
0    8                      Human development
1   11                                       
2    1                    Economic management
3    6  Social protection and risk management
4    5                  Trade and integration

In [26]:
themes_code.name.unique()

array(['Human development', '', 'Economic management',
       'Social protection and risk management', 'Trade and integration',
       'Public sector governance',
       'Environment and natural resources management',
       'Social dev/gender/inclusion',
       'Financial and private sector development', 'Rural development',
       'Urban development', 'Rule of law'], dtype=object)

In [27]:
themes_code.name.value_counts()

Environment and natural resources management    223
Rural development                               202
Human development                               197
Public sector governance                        184
Social protection and risk management           158
Financial and private sector development        130
                                                122
Social dev/gender/inclusion                     119
Trade and integration                            72
Urban development                                47
Economic management                              33
Rule of law                                      12
Name: name, dtype: int64

In [28]:
themes_code.code.unique()

array(['8', '11', '1', '6', '5', '2', '7', '4', '10', '9', '3'], dtype=object)

There are 122 empty espaces for the project names while there are no codes missing. Then our first setp is to match each code number to its corresponding project name.

In [29]:
#Creates dictionary with keys as code number and values as project names
code_name = {}
for code in themes_code.code.unique():
    i = 0
    while themes_code[themes_code.code == code].iloc[i,1] == '':
        i += 1        
    code_name[code] = code_name.get(code, themes_code[themes_code.code == code].iloc[i,1])
    
print(code_name)

{'8': 'Human development', '11': 'Environment and natural resources management', '1': 'Economic management', '6': 'Social protection and risk management', '5': 'Trade and integration', '2': 'Public sector governance', '7': 'Social dev/gender/inclusion', '4': 'Financial and private sector development', '10': 'Rural development', '9': 'Urban development', '3': 'Rule of law'}


In [30]:
#Creates a data frame pairing each code to its respective code name
df_code_name = \
    pd.DataFrame({'code':list(code_name.keys()), \
                  'name':list(code_name.values())}).set_index('code')

In [31]:
df_code_name

name
code                                              
8                                Human development
11    Environment and natural resources management
1                              Economic management
6            Social protection and risk management
5                            Trade and integration
2                         Public sector governance
7                      Social dev/gender/inclusion
4         Financial and private sector development
10                               Rural development
9                                Urban development
3                                      Rule of law

In [32]:
#Group themes_code by 'code' and use count() as aggregation function
#to get the number of times each code is present by sorting them
ranked_themes = themes_code.groupby('code').count()
#Rename 'name' column of ranked_themes to 'Number_of_Projects' 
ranked_themes = ranked_themes.rename(index=str, columns={'name':'total'})
#Join df_code_name to renaked_themes to have in the same table the project names and the codes
projects_rank = df_code_name.join(ranked_themes, how='inner').sort_values(by='total', ascending = False)
print('The 10 major projects themes are:\n\n', projects_rank.head(10))

The 10 major projects themes are:

                                               name  total
code                                                     
11    Environment and natural resources management    250
10                               Rural development    216
8                                Human development    210
2                         Public sector governance    199
6            Social protection and risk management    168
4         Financial and private sector development    146
7                      Social dev/gender/inclusion    130
5                            Trade and integration     77
9                                Urban development     50
1                              Economic management     38


### 3. In 2. above you will notice that some entries have only the code and the name is missing. Create a dataframe with the missing names filled in.

From 2. we have the data frame themes_code with 122 empty strings for the code names:

In [33]:
themes_code.name.value_counts()

Environment and natural resources management    223
Rural development                               202
Human development                               197
Public sector governance                        184
Social protection and risk management           158
Financial and private sector development        130
                                                122
Social dev/gender/inclusion                     119
Trade and integration                            72
Urban development                                47
Economic management                              33
Rule of law                                      12
Name: name, dtype: int64

In [34]:
themes_code.head()

code                                   name
0    8                      Human development
1   11                                       
2    1                    Economic management
3    6  Social protection and risk management
4    5                  Trade and integration

But we also created a dictionnary matching each code to its name:

In [35]:
print(code_name)

{'8': 'Human development', '11': 'Environment and natural resources management', '1': 'Economic management', '6': 'Social protection and risk management', '5': 'Trade and integration', '2': 'Public sector governance', '7': 'Social dev/gender/inclusion', '4': 'Financial and private sector development', '10': 'Rural development', '9': 'Urban development', '3': 'Rule of law'}


In [36]:
#Replaces empty strings in the name column by the proper name
j = 0 #counter to check the filled fields at the end matches the number of empty strings in the data frame 
for i in range(len(themes_code)):
    if themes_code.name.iloc[i] == '':
        themes_code.name.iloc[i] = code_name[themes_code.code.iloc[i]]
        j += 1

print('Number of fields filled: ', j)        

Number of fields filled:  122


In [37]:
themes_code.name.value_counts()

Environment and natural resources management    250
Rural development                               216
Human development                               210
Public sector governance                        199
Social protection and risk management           168
Financial and private sector development        146
Social dev/gender/inclusion                     130
Trade and integration                            77
Urban development                                50
Economic management                              38
Rule of law                                      15
Name: name, dtype: int64

There are no empty strings, so the data frame themes_code has been modified properly. We can check this too by printing the first rows of the data frame and checkin the name in the second row is filled: 

In [38]:
themes_code.head()

code                                          name
0    8                             Human development
1   11  Environment and natural resources management
2    1                           Economic management
3    6         Social protection and risk management
4    5                         Trade and integration